Migrate to Alpine

raqueleyeos opened the pull request.

afiestas committed:
Set ulimit to 1024

Openldap has a behaviour where it will allocate memory for each fd that
can be opened.

Docker does not set any ulimit nor propagate the default one (1024) so
the end result is an ulimit of 524288 making ldap use ~700MB of ram.

louis-sanchez committed:
Migrate LDAP to alpine linux

Dockerfile changes
· Change base image to alpine-node
· Delete monitor and hdb config files since we are now configuring all
DBs inside slapd.conf
· Delete bind_v2.ldif.  As far as I can tell this file is never being
used.

start.sh changes
· Delete add_ldif stuff since we are no longer adding schema with
ldapadd, but rather with slapd.conf
· add slapd.conf which will configure the LDAP for us at start time

NOTE: In our previous fedora image, we were configuring the LDAP using
the new slapd.d configuration (storing the config inside the LDAP
itself).  This is useful to be able to configure the LDAP at runtime
without having to restart it.  We have regressed back to the original
way of configuring the LDAP by using slapd.conf which is more intuitive,
but we lose the option of configuration without restarting the service.
openLDAP documentation indicates that at some point they will no
longer support slapd.conf, but it doesn't seem like that will happen
anytime soon.

louis-sanchez committed:
Use updateded build dependencies script in Dockerfile

Commit:
Dev variables for installing or removing and emove apk 'update'/'add' becouse it's executed by buildDependencies

louis-sanchez committed:
re-add add_ldif function to start.sh

· this function is needed as a fail-safe in case the ldapadd operation
fails because an entry already exists.  the error is ignored and the
script continues.

· the function was originally removed when migrating to alpine because
the -c flag should technically force a continue without breaking the
script, but this doesn't appear to be the case so we are putting it
back.

vHanda commented on Dockerfile:
```diff
@@ -1,28 +1,23 @@
+    /scripts-base/installExtraBuild.sh && \
+    apk add --no-cache openldap openldap-clients openldap-back-monitor && \
+    npm install -g --verbose eyeos-run-server eyeos-tags-to-dns eyeos-service-ready-notify-cli && \
     curl -L https://releases.hashicorp.com/serf/0.6.4/serf_0.6.4_linux_amd64.zip -o serf.zip && unzip serf.zip && mv serf /usr/bin/ && \
```
Serf is not required. It's present in the base image.

vHanda commented on start.sh:
```diff
@@ -2,29 +2,7 @@
-		fi
-	fi
-}
+cp /tmp/slapd.conf /etc/openldap/slapd.conf
```
Please just copy this in the correct location in the Dockerfile.

vHanda commented on Dockerfile:
```diff
@@ -1,28 +1,23 @@
-    /bin/cp /tmp/hdb /etc/openldap/slapd.d/cn=config/olcDatabase={2}hdb.ldif && \
-    /bin/cp /usr/share/openldap-servers/DB_CONFIG.example /var/lib/ldap/DB_CONFIG;chown -Rf ldap:ldap /var/lib/ldap/ && \
-    /usr/libexec/openldap/check-config.sh 
+    chmod 600 /tmp/ldap_rootpwd
```
Please see if these chmods are actually required? It seems that the files do have the correct permissions.

vHanda commented on Dockerfile:
```diff
@@ -1,28 +1,23 @@
 
-COPY files/monitor /tmp/
-COPY files/hdb /tmp/
 COPY files/base.ldif /tmp/
```
Please combine all these COPY statements into one command. It will further reduce the size of the docker image.

vHanda commented on Dockerfile:
```diff
@@ -1,28 +1,23 @@
-    /usr/libexec/openldap/check-config.sh 
+    chmod 600 /tmp/ldap_rootpwd
 
 CMD /tmp/start.sh
```
Lets not run files from /tmp.

/tmp should be cleared during the docker build process.

vHanda commented on Dockerfile:
```diff
@@ -1,28 +1,23 @@
+RUN apk update && \
+    /scripts-base/installExtraBuild.sh && \
+    apk add --no-cache openldap openldap-clients openldap-back-monitor && \
+    npm install -g --verbose eyeos-run-server eyeos-tags-to-dns eyeos-service-ready-notify-cli && \
```
Please clear the npm cache

kampde replied:
`eyeos-run-server` and `eyeos-tags-to-dns` are present in `alpine6-node-base` image. (`eyeos-service-ready-notify-cli` is not present in the base image, so it still needs to be installed).

vHanda commented:
@las560 - I've made some comments over here - https://github.com/Open365/docker-ldap-discovery/pull/2/commits/19d9a204f64b8fff79330e80195f260b4f4b6d00

Not all the comments are valid. Please address the ones that are.

kampde commented on Dockerfile:
```diff
@@ -1,28 +1,19 @@
-RUN mkdir -p /var/run/slapd && touch /var/run/slapd/slapd.pid && \
-    npm install -g eyeos-run-server eyeos-tags-to-dns eyeos-service-ready-notify-cli && \
+RUN /scripts-base/buildDependencies.sh --production --install && \
+    npm install -g --verbose eyeos-run-server eyeos-tags-to-dns eyeos-service-ready-notify-cli && \
```
`eyeos-run-server` and `eyeos-tags-to-dns` are present in the base image.

kampde commented on Dockerfile:
```diff
@@ -1,28 +1,19 @@
-    npm install -g eyeos-run-server eyeos-tags-to-dns eyeos-service-ready-notify-cli && \
+RUN /scripts-base/buildDependencies.sh --production --install && \
+    npm install -g --verbose eyeos-run-server eyeos-tags-to-dns eyeos-service-ready-notify-cli && \
     curl -L https://releases.hashicorp.com/serf/0.6.4/serf_0.6.4_linux_amd64.zip -o serf.zip && unzip serf.zip && mv serf /usr/bin/ && \
```
`serf` is present in the base image.

louis-sanchez committed:
Optimizations to Dockerfile

· move scripts and files to/var/service instead of /tmp.
· run npm clear cache
· remove installation of eyeos-run-server and eyeos-tags-to-dns as these
are already installed inside the alpine-base image
· change file paths to scripts that were located in /tmp to /var/service
inside start.sh

louis-sanchez committed:
Remove serf from Dockerfile

this is not necessary as serf is downloaded and installed inside the
alpine-node-base image

louis-sanchez commented:
@vHanda @kampde OK thanks.  I've addressed all of the relevant comments

kampde commented:
seems good to me.
